File: src/atcoder/abc141/abc141_e.go
package main

import (
	"fmt"
)

func ZAlgorithm(s string) []int {
	i := 1
	j := 0
	n := len(s)
	ans := make([]int, n)
	ans[0] = n

	for i < n {
		for i+j < n && s[j] == s[i+j] {
			j++
		}

		ans[i] = j

		if j == 0 {
			i++
			continue
		}

		k := 1

		for i+k < n && k+ans[k] < j {
			ans[i+k] = ans[k]
			k++
		}

		i += k
		j -= k
	}

	return ans
}

func Min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func ChMax(a *int, b int) {
	if *a < b {
		*a = b
	}
}

func main() {
	var n int
	var s string
	fmt.Scan(&n)
	fmt.Scan(&s)

	ans := 0

	for i := 0; i < n; i++ {
		cnt := ZAlgorithm(s[i:])

		for i, v := range cnt {
			ChMax(&ans, Min(i, v))
		}
	}
	fmt.Println(ans)
}

File: src/atcoder/abc141/abc141_e.rb
# Z algorithm
# sとs[i..]の共通接頭辞の長さの配列をO(|s|)で求める
#
# ```
# ZAlgorithm.new("abcbcba").solve # => [7,0,0,0,0,0,1]
# ```
class ZAlgorithm
  attr_reader :n, :s, :ans

  def initialize(s)
    @s = s
    @n = s.size
    @ans = Array.new(n, 0)
    ans[0] = n
  end

  def solve
    i = 1
    j = 0

    while i < n
      while i + j < n && s[j] == s[i + j]
        j += 1
      end

      ans[i] = j

      if j == 0
        i += 1
        next
      end

      k = 1

      while i + k < n && k + ans[k] < j
        ans[i + k] = ans[k]
        k += 1
      end

      i += k
      j -= k
    end

    ans
  end
end

n = gets.to_i
s = gets.chomp

ans = 0
n.times do |i|
  z = ZAlgorithm.new(s[i..]).solve
  cnt = z.zip(0..).map { |a, b| a < b ? a : b }.max
  ans = cnt if ans < cnt
end

pp ans
